### PR TITLE
Entity sound.

### PR DESCRIPTION
**Pull request by howard-stearns**
Initial crack at entity sound.

* There are a lot of changes here. In particular, how do I know I haven't broken anything about persistence or the wire protocol by adding a new entity property?

* There seems to be some sort of issue that causes the collision sounds to not be reliable. (Threading of either SoundCache or AudioInjector, or threading vs entityTree locking, a more general flakiness in sound, or all of these?) I have tried to make this so that the new feature can fail without breaking any existing features, such that we can work on the problem(s) in isolation as an improvement. But I have no idea if I've achieved that goal.

Outline:

1) There is a new entity property, collisionSoundURL, that should be editable in the Entity Properties tool, persistent, etc.  This is scattered through all of the files of this PR except for EntityTreeRenderer.

2) If an entity is subject to collisions and there is a non-empty collisionSoundURL, we attempt to compute the energy of the collision. If above a threshold, this is converted into a volume for the injected sound, which plays in another thread.  This is all concentrated in EntityTreeRenderer.  

(I realize that we don't try to load the sound until after the first collision, and thus it won't play that first time. I want to handle that as a later improvement.)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1a87ca04da50431a57f15cd7b53fef3a507e2384 available at: https://gist.github.com/55e4477a2000412937a6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4860.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4860.dmg

File: examples/html/entityProperties.html [hunk 1]
```diff
@@ -74,7 +74,7 @@
         function createEmitTextPropertyUpdateFunction(propertyName) {
             return function() {
                 var properties = {};
-                properties[propertyName] = this.value;
+		properties[propertyName] = this.value;
```
Comment (ZappoMan) [inline examples/html/entityProperties.html]: formatting looks broken here

File: libraries/entities/src/EntityPropertyFlags.h [hunk 2]
```diff
@@ -55,6 +55,7 @@ enum EntityPropertyList {
     PROP_DAMPING,
     PROP_LIFETIME,
     PROP_SCRIPT,
+    PROP_COLLISION_SOUND_URL,
```
Comment (ZappoMan) [inline libraries/entities/src/EntityPropertyFlags.h]: NO! You can't do this. All new properties must be added at the END.

Comment (ZappoMan) [inline libraries/entities/src/EntityPropertyFlags.h]: add this immediately after PROP_NAME

**Comment by ZappoMan** [comment]
this will definitely break the wire format as you have implemented it... you can't add new PROP_XXX in the middle of the EntityPropertyList enum.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 76e39fc95b21afbb8076ea5ddf8ec3cf6556f844 available at: https://gist.github.com/55e4477a2000412937a6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4860.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4860.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ef187b390ba784e6c6b98952c058562b537c8dd5 available at: https://gist.github.com/55e4477a2000412937a6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4860.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4860.dmg

File: examples/html/entityProperties.html [hunk 3]
```diff
@@ -244,6 +244,7 @@
             var elDensity = document.getElementById("property-density");
             var elIgnoreForCollisions = document.getElementById("property-ignore-for-collisions");
             var elCollisionsWillMove = document.getElementById("property-collisions-will-move");
+	    var elCollisionSoundURL = document.getElementById("property-collision-sound-url");
```
Comment (ZappoMan) [inline examples/html/entityProperties.html]: formatting still off here

File: examples/html/entityProperties.html [hunk 4]
```diff
@@ -433,6 +434,7 @@
                             elDensity.value = properties.density.toFixed(2);
                             elIgnoreForCollisions.checked = properties.ignoreForCollisions;
                             elCollisionsWillMove.checked = properties.collisionsWillMove;
+			    elCollisionSoundURL.value = properties.collisionSoundURL;
```
Comment (ZappoMan) [inline examples/html/entityProperties.html]: formatting still off here

File: examples/libraries/entityPropertyDialogBox.js [hunk 5]
```diff
@@ -185,6 +185,8 @@ EntityPropertyDialogBox = (function () {
         index++;
         array.push({ label: "Collisions Will Move:", type: "checkbox", value: properties.collisionsWillMove });
         index++;
+	array.push({ label: "Collision Sound URL:", value: properties.collisionSoundURL });
```
Comment (ZappoMan) [inline examples/libraries/entityPropertyDialogBox.js]: formatting still off here

**Comment by ZappoMan** [comment]
I tested the build locally and it seems to work. Sometimes I get no sound effect. Presumably when the tryLockForRead() fails to capture the lock and you bail without playing the sound

**Comment by hifi-gustavo** [comment]
Build console output for commit: b79a58c0c8312d0df5c3132ea9427f70483cbd0e available at: https://gist.github.com/55e4477a2000412937a6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4860.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4860.dmg

**Pull request merged**